### DISCUSSION TITLE
Ideas for more functions

### DISCUSSION THREAD

**Discussion by corasaurus-hex** [babashka/fs#56] [Ideas]
I came across [f.el](https://github.com/rejeep/f.el) and some of these functions look really useful for when you're messing around with the filesystem.

**Comment by corasaurus-hex** [comment]
A list of interesting functions that aren't exactly represented in the current codebase. The primitives are there to do most of these yourself but anything imo if it's sufficiently useful then we may as well save people from having to write it themselves.

* [f-split](https://github.com/rejeep/f.el#f-split-path)
* [f-filename](https://github.com/rejeep/f.el#f-filename-path)
* [f-common-parent](https://github.com/rejeep/f.el#f-common-parent-paths)
* [f-swap-ext](https://github.com/rejeep/f.el#f-swap-ext-path-ext)
* [f-base](https://github.com/rejeep/f.el#f-base-path)
* [f-relative](https://github.com/rejeep/f.el#f-relative-path-optional-dir)
* [f-slash](https://github.com/rejeep/f.el#f-slash-path)
* [f-append-bytes](https://github.com/rejeep/f.el#f-append-bytes-data-path) (and/or text?)
* [f-touch](https://github.com/rejeep/f.el#f-touch-path)
* [f-root?](https://github.com/rejeep/f.el#f-root-path)
* [f-parent-of?](https://github.com/rejeep/f.el#f-parent-of-path-a-path-b)
* [f-child-of?](https://github.com/rejeep/f.el#f-child-of-path-a-path-b)
* [f-ancestor-of?](https://github.com/rejeep/f.el#f-ancestor-of-path-a-path-b)
* [f-descendant-of?](https://github.com/rejeep/f.el#f-descendant-of-path-a-path-b)
* [f-empty?](https://github.com/rejeep/f.el#f-empty-path)
* [f-depth](https://github.com/rejeep/f.el#f-depth-path)
* [f-directories](https://github.com/rejeep/f.el#f-directories-path-optional-fn-recursive) (recursive, too)
* [f-files](https://github.com/rejeep/f.el#f-files-path-optional-fn-recursive) (recursive, too)
* [f-root](https://github.com/rejeep/f.el#f-root-)
* [f-traverse-upwards](https://github.com/rejeep/f.el#f-traverse-upwards-fn-optional-path) **!!!**

If I had to pick a few I'd say, in order of most to least desirability for me, are `f-traverse-upwards`, `f-common-parent`, and `f-slash`. I've needed all of these numerous times while scripting in the past.

**Reply by borkdude** [reply]
Some interesting ones in there! When we need one of these, we can take a look how they did it here.

**Comment by jasalt** [comment]
Another idea would be to have `locate-dominating-file` as in Emacs (given filename, recursing up until finding it):

https://github.com/emacs-mirror/emacs/blob/bc4c5818dfb3c16f817514d3ae3bc074b4c4c007/lisp/files.el#L1187

Seems something like that could be composed with [f-traverse-upwards](https://github.com/rejeep/f.el#f-traverse-upwards-fn-optional-path) from previous comment also. Wondering which would fit neatly with this library.
